src/SonarQube.Client/Services/SonarQubeService.cs

```diff
@@ -0,0 +1,253 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using SonarQube.Client.Models;
+using SonarQube.Client.Services;
+
+namespace SonarQube.Client
+{
+    public class SonarQubeService
+    {
+        internal const int MaximumPageSize = 500;
+        internal readonly Version OrganizationsFeatureMinimalVersion = new Version(6, 2);
+
+        private readonly ISonarQubeClient sonarqubeClient;
+
+        private ConnectionDTO connection;
+        private bool isConnected;
+        private Version serverVersion;
+
+        public SonarQubeService(ISonarQubeClient sonarqubeClient)
+        {
+            this.sonarqubeClient = sonarqubeClient;
+        }
+
+        public bool HasOrganizationsFeature
+        {
+            get
+            {
+                if (!this.isConnected)
+                {
+                    throw new InvalidOperationException("This operation expects the service to be connected.");
+                }
+
+                return this.serverVersion >= OrganizationsFeatureMinimalVersion;
+            }
+        }
+
+        public async Task ConnectAsync(Connection connection, CancellationToken token)
+        {
+            if (this.isConnected)
+            {
+                throw new InvalidOperationException("This operation expects the service to be connected.");
+            }
+
+            var connectionDto = new ConnectionDTO
+            {
+                Authentication = connection.Authentication,
+                ServerUri = connection.ServerUri,
+                Login = connection.Login,
+                Password = connection.Password
+            };
+
+            var credentialsValidationResult = await this.sonarqubeClient.ValidateCredentialsAsync(connectionDto, token);
+            if (credentialsValidationResult.IsFailure)
+            {
+
+                // TODO: something went wrong
+            }
+            if (!credentialsValidationResult.Value.AreValid)
+            {
+                // TODO: credentials are not valid
+            }
+
+            var versionResult = await this.sonarqubeClient.GetVersionAsync(connectionDto, token);
+            if (versionResult.IsFailure)
+            {
+                // TODO: something went wrong
+            }
+
+            Version version;
+            if (!Version.TryParse(versionResult.Value.Version, out version))
+            {
+                // TODO: version is not valid
+            }
+
+            this.connection = connectionDto;
+            this.serverVersion = version;
+            this.isConnected = true;
+        }
+        public async Task<IList<Organization>> GetAllOrganizationsAsync(CancellationToken token)
+        {
+            if (!this.isConnected)
+            {
+                throw new InvalidOperationException("This operation expects the service to be connected.");
+            }
```

Comment:
Could factor this check out to helper method - simpler if the error message is ever changed.

Reply:
Done.